modules/core/ShaderShadow.cpp

```diff
@@ -0,0 +1,44 @@
+#include "ShaderShadow.h"
+#include "Scene.h"
+#include "Ray.h"
+
+namespace rt {
+	Vec3f CShaderShadow::shade(const Ray& ray) const
+	{
+		// Traverse ray behind the geometry to make it fully transparent
+		Ray R(ray.hitPoint(), ray.dir, ray.counter);	// traverse ray
+		Vec3f res = R.reTrace(m_scene);
+		
+		// Gathering shadows
+		Vec3f shadingNormal = ray.hit->getShadingNormal(ray);
+		
+		Ray I(ray.hitPoint(shadingNormal));				// shadow ray
+		Vec3f L_possible = Vec3f::all(0);
```

Comment:
I'm guessing the value of L_actual is incremented where there's an actual geometry between the ray and the light source (that's why we check if the light has the shadow flag enabled and if it intersects any object), but I can't wrap my head around why L_possible is needed. Would you mind helping me understand this?

Reply:
`L_actual` is incremented only in the case when the shading point is not in shadow, whereas `L_possible` is incremented whenever light theoretically can reach the point (if there is no other geometry in the scene). 
The expression 

> `!pLight->shadow() || !m_scene.if_intersect(I)`

 is the same as 

> `pLight->shadow() && m_scene.if_intersect(I)`

meaning that  _when the light source casts shadows and the point is occluded from that light source_.

We need both values because the ratio `L_actual / L_possible` characterize how _dark_ or how _deep_ the shadow is. This value likes between 0 and 1, so we can use it to attenuate what we see through the object to emulate shadow in transparent surface.